src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -443,8 +444,15 @@ ArrayDeque<ParallelBlockStart> leastCommonAncestor(@NonNull final Set<FlowNode>
             throw new IllegalStateException("No least common ancestor found from " + heads);
         }
 
-        // If we hit issues with the ordering of blocks by depth, apply a sorting to the parallels by depth
-        return convertForksToBlockStarts(parallelForks);
+        // The result must be in reverse topological order, i.e. inner branches must be visited before outer branches.
+        return convertForksToBlockStarts(parallelForks).stream().sorted((pbs1, pbs2) -> {
+            if (pbs1.forkStart.getEnclosingBlocks().contains(pbs2.forkStart)) {
+                return -1;
+            } else if (pbs2.forkStart.getEnclosingBlocks().contains(pbs1.forkStart)) {
+                return 1;
+            }
+            return 0;
```

Comment:
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Comparator.html this is not “consistent with equals”. That is generally harmless when sorting a list (and I suppose also a stream) but I do not think you can reliably implement a topological sort this way. For example in `jshell`

```java
import java.util.stream.*;
Stream.of("axx a ax".split(" ")).sorted((x, y) -> x.startsWith(y) ? 1 : y.startsWith(x) ? -1 : 0).collect(Collectors.joining(" "))
```

works (`a ax axx`) but on the input `axx b a ax` you get back the same list, whereas you might expect `a b ax axx` or similar.

Whether there is any observable bug in _this_ function as a result is another question.

Reply:
Hmm yes, I can look into this.

> Whether there is any observable bug in this function as a result is another question.

Perhaps, if you have more than 3 levels of nested parallel steps, or if you have multiple parallel step children under a single parallel step, although I wonder if the latter case would work regardless of this change because of other issues.

Reply:
From some quick testing, I have only been able to reproduce cases where the outermost parallel step is the second to last element before the sort, when it needs to always be the final element. Other than that, the pre-sort order is fine. Perhaps a more targeted fix could be made further up in the method, and it seems plausible that with some really unusual Pipelines there is some reachable case that I was not able to trigger, but I am not sure.

Reply:
Would recommend at least leaving in a code comment warning that this logic would not work for exotic cases which may or may not be reachable.

Reply:
See https://github.com/jenkinsci/workflow-api-plugin/pull/288.